internal/pkg/deploy/sns.go

```diff
@@ -0,0 +1,89 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package deploy holds the structures to deploy infrastructure resources.
+// This file defines workload deployment resources.
+package deploy
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws/arn"
+)
+
+var errInvalidTopicARN = errors.New("ARN is not a Copilot SNS topic")
+var errInvalidARN = errors.New("invalid ARN format")
+var errInvalidARNService = errors.New("ARN is not an SNS topic")
+var errInvalidComponent = errors.New("Topic must be initialized with valid app, env, and workload")
+
+// Topic holds information about a Copilot SNS topic and its ARN, ID, and Name.
+type Topic struct {
+	awsARN arn.ARN
+	prefix string
+	wkld   string
+
+	name string
+}
+
+// NewTopic creates a new Topic struct, validating the ARN as a Copilot-managed SNS topic.
+// This function will
+func NewTopic(inputARN string, app, env, wkld string) (*Topic, error) {
+	if app == "" || env == "" || wkld == "" {
+		return nil, errInvalidComponent
+	}
+
+	parsedARN, err := arn.Parse(inputARN)
+	if err != nil {
+		return nil, errInvalidARN
+	}
+
+	if parsedARN.Service != snsServiceName {
+		return nil, errInvalidARNService
+	}
+
+	t := &Topic{
+		prefix: fmt.Sprintf(fmtSNSTopicNamePrefix, app, env, wkld),
+		wkld:   wkld,
+		awsARN: parsedARN,
+	}
+
+	if err = t.validateAndExtractName(); err != nil {
+		return nil, err
+	}
+
+	return t, nil
+}
+
+// Name returns the name of the given SNS topic, stripped of its "app-env-wkld" prefix.
+func (t Topic) Name() string {
+	return t.name
+}
+
+// ARN returns the full ARN of the SNS topic.
+func (t Topic) ARN() string {
```

Comment:
What do you think of having only two public methods:
```go
func (t Topic) ARN() string { return t.arn.String() }
func (t Topic) String() string { return fmt.Sprintf("%s (%s)", t.name, t.workload) 
```
This way in the selector you can use `t.String()`